src/tiled/commanddialog.cpp

```diff
@@ -72,23 +80,63 @@ void CommandDialog::closeEvent(QCloseEvent *event)
 void CommandDialog::setShortcut(const QKeySequence &keySequence)
 {
     const QModelIndex &current = mUi->treeView->currentIndex();
-    if (current.row() < mUi->treeView->model()->rowCount(QModelIndex()))
+    if (current.row() < mUi->treeView->model()->rowCount())
         mUi->treeView->model()->setShortcut(current, keySequence);
 }
 
-void CommandDialog::updateKeySequenceEdit(const QModelIndex &current, const QModelIndex &)
+void CommandDialog::setSaveBeforeExecute(int state)
+{
+    const QModelIndex &current = mUi->treeView->currentIndex();
+    if (current.row() < mUi->treeView->model()->rowCount())
+        mUi->treeView->model()->setSaveBeforeExecute(current, state);
+}
+
+void CommandDialog::setCommand(const QString &text)
+{
+    const QModelIndex &current = mUi->treeView->currentIndex();
+    if (current.row() < mUi->treeView->model()->rowCount())
+        mUi->treeView->model()->setCommand(current, text);
+}
+
+void CommandDialog::updateWidgets(const QModelIndex &current, const QModelIndex &)
 {
-    if (current.row() < mUi->treeView->model()->rowCount(QModelIndex()) - 1) {
-        mUi->keySequenceEdit->setEnabled(true);
-        mUi->clearButton->setEnabled(true);
+    bool enable = true;
+
+    if (current.row() < mUi->treeView->model()->rowCount() - 1) {
+        mUi->commandEdit->setEnabled(enable);
+        mUi->browseButton->setEnabled(enable);
+        mUi->keySequenceEdit->setEnabled(enable);
+        mUi->clearButton->setEnabled(enable);
+        mUi->saveBox->setEnabled(enable);
+
         mUi->keySequenceEdit->setKeySequence(mUi->treeView->model()->shortcut(current));
-    } else {
+        mUi->saveBox->setChecked(mUi->treeView->model()->saveBeforeExecute(current));
+        mUi->commandEdit->setText(mUi->treeView->model()->command(current));
+    }
+    else {
+        enable = false;
+
+        mUi->commandEdit->clear();
         mUi->keySequenceEdit->clear();
-        mUi->keySequenceEdit->setEnabled(false);
-        mUi->clearButton->setEnabled(false);
+        mUi->saveBox->setChecked(enable);
+        mUi->commandEdit->setEnabled(enable);
+        mUi->browseButton->setEnabled(enable);
+        mUi->keySequenceEdit->setEnabled(enable);
+        mUi->clearButton->setEnabled(enable);
+        mUi->saveBox->setEnabled(enable);
```

Comment:
You know you still need to de-duplicate these lines, right? I mean now you're using a variable, but...